src/IR/exprs.ts

```diff
@@ -123,14 +114,113 @@ export function keyValue(key: Expr, value: Expr): KeyValue {
   };
 }
 
-export function polygolfOp(op: OpCode, ...args: Expr[]): PolygolfOp {
+function _polygolfOp(op: OpCode, ...args: Expr[]): PolygolfOp {
   return {
     kind: "PolygolfOp",
     op,
     args,
   };
 }
 
+export function polygolfOp(op: OpCode, ...args: Expr[]): Expr {
+  if (op === "neg") {
+    return polygolfOp("mul", int(-1), args[0]);
+  }
+  if (op === "sub") {
+    return polygolfOp("add", args[0], polygolfOp("neg", args[1]));
+  }
+  if (isBinary(op) && isAssociative(op)) {
+    args = args.flatMap((x) =>
+      x.kind === "PolygolfOp" && x.op === op ? x.args : [x]
+    );
+    if (op === "add") args = simplifyPolynomial(args);
+    else {
+      if (isCommutative(op)) {
+        args.sort(compareTerms);
```

Comment:
It's not clear that the `compareTerms` is correct (it's not anti-symmetric for example). Since this seems to just be putting integer literals at the start, why not something like (untested, and you'll need to also remove the definition of `compareTerms`):
```suggestion
        args = args.filter(x => x.kind === "IntegerLiteral").concat(args.filter(x => x.kind !== "IntegerLiteral"));
```